base/sparse/symmetric.jl

```diff
@@ -0,0 +1,72 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+function Symmetric(A::SparseMatrixCSC, uplo::Symbol=:U)
+    checksquare(A)
+    Symmetric{eltype(A), typeof(A)}(A, Base.LinAlg.char_uplo(uplo))  # preserve A
+end
+
+(*)(A::Symmetric{TA,SparseMatrixCSC{TA,S}}, x::StridedVecOrMat{Tx}) where {TA,S,Tx} = A_mul_B(A, x)
+
+function A_mul_B!(α::Number, A::Symmetric{TA,SparseMatrixCSC{TA,S}}, B::StridedVecOrMat, β::Number, C::StridedVecOrMat) where {TA,S}
+    A.data.n == size(B, 1) || throw(DimensionMismatch())
+    A.data.m == size(C, 1) || throw(DimensionMismatch())
+    A.uplo == 'U' ? A_mul_B_U_kernel!(α, A, B, β, C) : A_mul_B_L_kernel!(α, A, B, β, C)
+end
+
+function A_mul_B(A::Symmetric{TA,SparseMatrixCSC{TA,S}}, x::StridedVector{Tx}) where {TA,S,Tx}
+    T = promote_type(TA, Tx)
+    A_mul_B!(one(T), A, x, zero(T), similar(x, T, A.data.n))
+end
+
+function A_mul_B(A::Symmetric{TA,SparseMatrixCSC{TA,S}}, B::StridedMatrix{Tx}) where {TA,S,Tx}
+    T = promote_type(TA, Tx)
+    A_mul_B!(one(T), A, B, zero(T), similar(B, T, (A.data.n, size(B, 2))))
+end
+
+function A_mul_B_U_kernel!(α::Number, A::Symmetric{TA,SparseMatrixCSC{TA,S}}, B::StridedVecOrMat, β::Number, C::StridedVecOrMat) where {TA,S}
```

Comment:
Using the matrix in my gist, here's the benchmark of `A * b`:
```
BenchmarkTools.Trial: 
  memory estimate:  3.85 MiB
  allocs estimate:  2
  --------------
  minimum time:     21.475 ms (0.00% GC)
  median time:      22.218 ms (0.00% GC)
  mean time:        22.425 ms (1.17% GC)
  maximum time:     27.263 ms (5.92% GC)
  --------------
  samples:          223
  evals/sample:     1
```
Now `U * b` and `L * b` using separate kernels:
```
BenchmarkTools.Trial: 
  memory estimate:  3.85 MiB
  allocs estimate:  2
  --------------
  minimum time:     28.259 ms (0.00% GC)
  median time:      29.112 ms (0.00% GC)
  mean time:        29.481 ms (0.88% GC)
  maximum time:     35.077 ms (0.00% GC)
  --------------
  samples:          170
  evals/sample:     1
BenchmarkTools.Trial: 
  memory estimate:  3.85 MiB
  allocs estimate:  2
  --------------
  minimum time:     33.965 ms (0.00% GC)
  median time:      35.219 ms (0.00% GC)
  mean time:        35.797 ms (0.92% GC)
  maximum time:     40.802 ms (0.00% GC)
  --------------
  samples:          140
  evals/sample:     1
```
and `U * b` and `L * b` using the joint kernel:
```
BenchmarkTools.Trial: 
  memory estimate:  3.85 MiB
  allocs estimate:  4
  --------------
  minimum time:     33.787 ms (0.00% GC)
  median time:      34.744 ms (0.00% GC)
  mean time:        35.347 ms (0.88% GC)
  maximum time:     40.835 ms (0.00% GC)
  --------------
  samples:          142
  evals/sample:     1
BenchmarkTools.Trial: 
  memory estimate:  3.85 MiB
  allocs estimate:  4
  --------------
  minimum time:     33.890 ms (0.00% GC)
  median time:      34.753 ms (0.00% GC)
  mean time:        35.240 ms (0.89% GC)
  maximum time:     41.668 ms (0.00% GC)
  --------------
  samples:          142
  evals/sample:     1
```
I didn't know about `Val`. It improves over my initial joint version somewhat, but there'll still be a noticeable difference over large amounts of matrix-vector products.